### DISCUSSION TITLE
JSON object in PostGIS function comment not generating the TileJSON as expected

### DISCUSSION THREAD

**Discussion by jjcfrancisco** [maplibre/martin#1034] [Q&A]
Hello,

I have a database function that pulls data from different tables and I've set it in the `config.yaml`. One column in such function is `uid` which I want to set as the primary id... I think in the tables sources in the `config.yaml` is known as `id_column`. As far as I can see, the only way to set the primary id in a function in martin is by adding a comment to the database function. The issue I have is that the comment does not get picked up by martin. As a test, I tried setting the description of the function but this does not seem to work either.

This is what I tried:

```sql
DO $do$ BEGIN
    EXECUTE 'COMMENT ON FUNCTION gis_functions.get_project(integer, integer, integer, json) IS $tj$' || $$
    {
        "description": "my test description"
        "id": "uid"
    }
    $$::json || '$tj$';
END $do$;
```

This is the generated tilejson

```json
{
  "tiles": {
    "project": {
      "content_type": "application/x-protobuf",
      "description": "gis_functions.get_project"
    }
  },
  "sprites": {},
  "fonts": {}
}
```

**Comment by nyurik** [comment]
@jjcfrancisco I am not sure I understand the issue (we might need to improve documentation? If so, any suggestions are welcome):
* the json you posted is NOT tilejson.  Its the catalog of all available sources, sprites, and fonts. You got it from `/catalog` URL. To get tilejson you should use `/project` instead (that's the name of your source).
* the feature ID (part of the MVT tile) is generated by your own function when it calls [ST_AsMVT](https://postgis.net/docs/ST_AsMVT.html) -- use the last variant `bytea ST_AsMVT(anyelement row, text name, integer extent, text geom_name, text feature_id_name)` -- the last parameter is the name of the column to use as a feature id.
* ID is not part of the tilejson afaik, so you don't need to specify that the internal `uid` table column is being used as the id.
* Note that if you were to let Martin auto-configure **table** sources (rather than a function source), you would specify that `uid` column of a table should be used as a feature id.

**Reply by jjcfrancisco** [reply]
@nyurik thanks for the clarification. I managed to get the feature id showing now - such a simple tweak. It'd be good make a quick and simple reference to this, just like you explained here.

The one that I am still scratching my head is how do I modify a description. This does not work (taken from documentation):

```sql
CREATE OR REPLACE
    FUNCTION function_zxy_query(z integer, x integer, y integer, query_params json)
    RETURNS bytea AS $$
DECLARE
  mvt bytea;
BEGIN
  SELECT INTO mvt ST_AsMVT(tile, 'function_zxy_query', 4096, 'geom') FROM (
    SELECT
      ST_AsMVTGeom(
          ST_Transform(ST_CurveToLine(geom), 3857),
          ST_TileEnvelope(z, x, y),
          4096, 64, true) AS geom
    FROM table_source
    WHERE geom && ST_Transform(ST_TileEnvelope(z, x, y), 4326)
  ) as tile WHERE geom IS NOT NULL;

  RETURN mvt;
END
$$ LANGUAGE plpgsql IMMUTABLE STRICT PARALLEL SAFE;

DO $do$ BEGIN
    EXECUTE 'COMMENT ON FUNCTION function_zxy_query(INT4, INT4, INT4, json) IS $tj$' || $$
    {
        "description": "my new description",
        "attribution": "my attribution",
        "vector_layers": [
            {
                "id": "my_layer_id",
                "fields": {
                    "field1": "String",
                    "field2": "Number"
                }
            }
        ]
    }
    $$::json || '$tj$';
END $do$;
```

tilejson from `test`:
```json
{"tiles":{"test":{"content_type":"application/x-protobuf","description":"public.function_zxy_query"}},"sprites":{},"fonts":{}}
```

**Reply by nyurik** [reply]
@jjcfrancisco once again, you are not copy/pasting `tilejson`, you are copying the catalog.

**Reply by nyurik** [reply]
also, please specify the exact version you are using, as shown with `martin --version`

**Reply by jjcfrancisco** [reply]
Apologies, this is the tilejson:
```json
{"tilejson":"3.0.0","tiles":["http://0.0.0.0:3001/test/{z}/{x}/{y}"],"description":"public.function_zxy_query","name":"test"}
```

As for version, just latest `v0.11.1`

**Reply by nyurik** [reply]
thx @jjcfrancisco , i would have to dig deeper into it.  We have a test for this in https://github.com/maplibre/martin/blob/2aa736f8af01ca494420d1a8f75e879bc6d5a44f/tests/fixtures/functions/function_zxy_query.sql -- which has been working so far, so try to see if there is anything different with your usecase?

**Reply by jjcfrancisco** [reply]
Thanks @nyurik. I tried the example as well but does not work. For reference it was a PG15 + PostGIS 3.3. Thanks again for looking into, nothing to worry about as this is a minor thing in my opinion.

**Reply by nyurik** [reply]
Fixed in v0.11.3

**Reply by jjcfrancisco** [reply]
Thanks @nyurik so fast 🚀
